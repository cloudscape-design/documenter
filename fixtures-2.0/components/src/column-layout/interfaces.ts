// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { BaseComponentProps } from '../internal/base-component';
import React from 'react';
import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
import { ColumnLayoutBreakpoint } from './internal';

export interface ColumnLayoutProps extends BaseComponentProps {
  /**
   * Specifies the number of columns in each grid row.
   * Valid values are any integer between 1 and 4.
   */
  columns?: number;

  /**
   * Specifies the content type. This determines the spacing of the grid.
   */
  variant?: ColumnLayoutProps.Variant;

  /**
   * Controls whether dividers are placed between rows and columns.
   */
  borders?: ColumnLayoutProps.Borders;

  /**
   * Determines whether the default gutters between columns are removed.
   */
  disableGutters?: boolean;

  /**
   * The columns to render.
   */
  children?: React.ReactNode;
}

export namespace ColumnLayoutProps {
  export type Variant = 'default' | 'text-grid';
  export type Borders = 'none' | 'vertical' | 'horizontal' | 'all';
}

export interface InternalColumnLayoutProps extends ColumnLayoutProps, InternalBaseComponentProps {
  __breakpoint?: ColumnLayoutBreakpoint;
}
