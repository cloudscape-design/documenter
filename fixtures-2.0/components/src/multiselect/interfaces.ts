// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { OptionDefinition, OptionGroup as OptionGroupDefinition } from '../internal/components/option/interfaces';
import { NonCancelableEventHandler } from '../internal/events';

import { BaseSelectProps } from '../select/interfaces';

/*
 * HACK: Cast the component to a named parametrized interface.
 *
 * This lets us use React.forwardRef and still let the component have type
 * parameters, and the naming convention lets the documenter know that this is
 * a forwardRef-wrapped component.
 *
 * We don't need to expose this type to customers because it's just a simple
 * function type.
 */
export interface MultiselectForwardRefType {
  (props: MultiselectProps & { ref?: React.Ref<MultiselectProps.Ref> }): JSX.Element;
}

export interface MultiselectProps extends BaseSelectProps {
  /**
   * Specifies the currently selected options.
   * Provide an empty array to clear the selection.
   */
  selectedOptions: ReadonlyArray<MultiselectProps.Option>;
  /**
   * Determines whether the dropdown list stays open after the user selects an item.
   */
  keepOpen?: boolean;
  /**
   * Specifies the maximum number of displayed tokens. If the property isn't set, all of the tokens are displayed.
   */
  tokenLimit?: number;
  /**
   * Hides the tokens displayed underneath the component.
   * Only use this if the selected options are displayed elsewhere on the page.
   */
  hideTokens?: boolean;
  /**
   * Specifies an `aria-label` for the token deselection button.
   */
  deselectAriaLabel?: MultiselectProps.DeselectAriaLabelFunction;
  /**
   * An object containing all the localized strings required by the component.
   * Note that the string for `tokenLimitShowMore` should not contain the number of hidden tokens
   * because it will be added by the component automatically.
   */
  i18nStrings?: MultiselectProps.I18nStrings;
  /**
   * Called when the user selects or deselects an option.
   * The event `detail` contains the current `selectedOptions`.
   */
  onChange?: NonCancelableEventHandler<MultiselectProps.MultiselectChangeDetail>;
}

export namespace MultiselectProps {
  export type Option = OptionDefinition;
  export type OptionGroup = OptionGroupDefinition;
  export type Options = ReadonlyArray<Option | OptionGroup>;
  export type DeselectAriaLabelFunction = (option: Option) => string;

  export interface I18nStrings {
    tokenLimitShowFewer?: string;
    tokenLimitShowMore?: string;
  }

  export interface MultiselectChangeDetail {
    selectedOptions: ReadonlyArray<Option>;
  }

  export interface Ref {
    /**
     * Sets focus on the element without opening the dropdown or showing a visual focus indicator.
     */
    focus(): void;
  }
}
