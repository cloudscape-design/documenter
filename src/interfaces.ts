// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// Component interfaces

export interface ComponentDefinition {
  name: string;
  releaseStatus: string;
  version?: string;
  description?: string;
  properties: ComponentProperty[];
  regions: ComponentRegion[];
  functions: ComponentFunction[];
  events: EventHandler[];
}

export interface ComponentProperty {
  name: string;
  description?: string;
  optional: boolean;
  type: string;
  inlineType?: InlineTypeDefinition;
  defaultValue?: string;
  deprecatedTag?: string;
  visualRefreshTag?: string;
  i18nTag?: true;
}

export interface ComponentRegion {
  name: string;
  isDefault: boolean;
  displayName?: string;
  description?: string;
  deprecatedTag: string | undefined;
  inlineType?: InlineTypeDefinition;
  visualRefreshTag: string | undefined;
  i18nTag?: true;
}

export interface ComponentFunction {
  name: string;
  description?: string;
  parameters: FunctionParameter[];
  returnType: string;
}

export type InlineTypeDefinition = ObjectDefinition | FunctionDefinition | UnionTypeDefinition;

export interface ObjectDefinition {
  name: string;
  type: 'object';
  properties: ObjectDefinitionProperty[];
}

export interface ObjectDefinitionProperty {
  name: string;
  optional: boolean;
  type: string;
}

export interface FunctionDefinition {
  name: string;
  type: 'function';
  returnType: string;
  parameters: FunctionParameter[];
}

export interface FunctionParameter {
  name: string;
  type: string;
}

export interface UnionTypeDefinition {
  name: string;
  type: 'union';
  values: string[];
}

export interface EventHandler {
  name: string;
  description?: string;
  detailType?: string;
  detailInlineType?: InlineTypeDefinition;
  cancelable: boolean;
  deprecatedTag?: string;
}

// Test utils interfaces

export interface TestUtilParameter {
  name: string;
  typeName?: string;
  description?: string;
  flags: { isOptional?: boolean };
  defaultValue?: string;
}

export interface TestUtilMethod {
  name: string;
  description?: string;
  returnType?: string;
  parameters: Array<TestUtilParameter>;
  inheritedFrom?: {
    name: string;
  };
}

export interface TestUtilsDoc {
  name: string;
  methods: Array<TestUtilMethod>;
}
