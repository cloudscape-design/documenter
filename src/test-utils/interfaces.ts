// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export interface Parameter {
  name: string;
  typeName?: string;
  description?: string;
  flags: { isOptional?: boolean };
  defaultValue?: string;
}

interface TypeArgument {
  name: string;
}

export interface TestUtilMethod {
  name: string;
  description?: string;
  returnType?: {
    name: string;
    isNullable: boolean;
    typeArguments?: Array<TypeArgument>;
  };
  parameters: Array<Parameter>;
  inheritedFrom?: {
    name: string;
  };
}

export interface TestUtilsDoc {
  name: string;
  methods: Array<TestUtilMethod>;
}

export interface TestUtilsDefinition {
  classes: Array<TestUtilsDoc>;
}
