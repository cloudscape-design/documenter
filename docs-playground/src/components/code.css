.code {
  background: transparent;
  margin: 0;
  padding: 0;
  font-family: monospace;
  font-size: 1em;
}

.code-token-link {
  font-family: monospace;
  font-size: 1em;
}

.code-token-symbol {
  /* $color-text-body-secondary */
  color: #545b64;
}
