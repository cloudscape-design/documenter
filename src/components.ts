// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as path from 'node:path';
import * as assert from 'node:assert/strict';
import ts from 'typescript';
import { glob } from 'fast-glob';
import { pascalCase } from 'change-case';

import {
  ComponentDefinition,
  ComponentFunction,
  ComponentProperty,
  ComponentRegion,
  EventHandler,
  InlineTypeDefinition,
} from './interfaces';

import {
  nodeToString,
  typeToString,
  joinSymbolDisplayParts,
  isTruthy,
  sortSymbolsByName,
  signatureDeclarationToString,
} from './util';

const EVENT_HANDLER_REGEX = /^on[A-Z]/;
const PASSTHROUGH_TYPE_NAMES = ['HTMLElement'];

export function documentComponents(tsconfigPath: string, publicFilesGlob: string): ComponentDefinition[] {
  // HACK: assuming "src" directory cause it's easier for now
  const allFiles = glob.sync(path.join(path.dirname(tsconfigPath), 'src/**/*.{ts,tsx}'), { absolute: true });
  const componentEntryFiles = glob.sync(publicFilesGlob, { absolute: true });

  // Don't provide the tsconfig because it unpredictably alters how TypeScript parses source files.
  // Thankfully, we don't really need much config-specific stuff here.
  const program = ts.createProgram(allFiles, { strict: true });

  return program
    .getSourceFiles()
    .filter(({ fileName }) => componentEntryFiles.includes(fileName))
    .flatMap(sourceFile => buildComponentDefinition(program, sourceFile))
    .filter(isTruthy);
}

export function buildComponentDefinition(
  program: ts.Program,
  sourceFile: ts.SourceFile,
): ComponentDefinition | undefined {
  /** The expected component name, based on the naming convention of the folder. */
  const componentName = pascalCase(path.basename(path.dirname(sourceFile.fileName)));

  /** The expected props type name, based on the naming convention of the folder. */
  const propsName = `${componentName}Props`;

  /** All the types and values exported from this source file. */
  const fileExports = program.getTypeChecker().getSymbolAtLocation(sourceFile)?.exports;
  if (!fileExports) {
    console.warn(`File has no exports! ${sourceFile.fileName}`);
    return;
  }

  /**
   * The AST node where the component was most likely defined.
   * This node's parameters can inspected to fetch default property values.
   */
  let componentNode: ts.SignatureDeclaration | undefined;

  /** The likely interface/type of the component properties. */
  let propsType: ts.Type | undefined;

  fileExports.forEach((tsSymbol, exportedName) => {
    if (exportedName === 'default') {
      // Get the non-type declaration that matches this export.
      // It's not possible to have two default exports, so only one is a non-type node.
      const node = tsSymbol.getDeclarations()?.find(node => !ts.isTypeNode(node));
      componentNode = node && getComponentFunctionNode(program, node);
    }

    if (exportedName === propsName) {
      propsType = program.getTypeChecker().getDeclaredTypeOfSymbol(tsSymbol);
    }
  });

  assert.ok(componentNode, `File has no default non-type export: ${sourceFile.fileName}`);
  assert.ok(propsType, `File has no props export (${propsName}): ${sourceFile.fileName}`);

  //
  // Get component documentation tags
  //

  let componentVersion: string | undefined;
  let componentDoc: string | undefined;
  let componentBeta = false;

  const componentSymbol = program.getTypeChecker().getSymbolAtLocation(componentNode.name ?? componentNode);
  if (componentSymbol) {
    const versionTag = componentSymbol.getJsDocTags().find(tag => tag.name === 'version');
    componentVersion = versionTag?.text && joinSymbolDisplayParts(versionTag.text);
    componentDoc =
      joinSymbolDisplayParts(componentSymbol.getDocumentationComment(program.getTypeChecker())) || undefined;
    componentBeta = componentSymbol.getJsDocTags().some(tag => tag.name === 'beta');
  }

  //
  // Extract default values from componentNode
  //

  const defaultValues = new Map<string, string>();

  if (componentNode.parameters[0] && ts.isObjectBindingPattern(componentNode.parameters[0].name)) {
    for (const bindingElement of componentNode.parameters[0].name.elements) {
      if (ts.isIdentifier(bindingElement.name) && bindingElement.initializer) {
        const defaultValueString = nodeToString(sourceFile, bindingElement.initializer, ts.EmitHint.Expression);
        defaultValues.set(bindingElement.name.getText(), defaultValueString);
      }
    }
  }

  //
  // Find a type under the props type named "Ref" and pull the functions from there.
  //
  const functions: Array<ComponentFunction> = [];

  const refSymbol = propsType.getSymbol()?.exports?.get(ts.escapeLeadingUnderscores('Ref'));
  const refProps = refSymbol && program.getTypeChecker().getDeclaredTypeOfSymbol(refSymbol).getProperties();
  for (const fnSymbol of refProps ?? []) {
    const refFunctionType = program.getTypeChecker().getTypeOfSymbol(fnSymbol);
    // stripUnionWithUndefined because optional functions are treated as being unioned with undefined.
    // FIXME: Should ref functions even be optional? We don't document them as optional except in typings.
    const funcSignature = stripUnionWithUndefined(refFunctionType).getCallSignatures()[0];
    assert.ok(funcSignature, `Could not find function signature for ${fnSymbol.getName()}`);

    functions.push({
      name: fnSymbol.getName(),
      description: joinSymbolDisplayParts(fnSymbol.getDocumentationComment(program.getTypeChecker())),
      parameters: funcSignature.parameters.map(parameter => ({
        name: parameter.getName(),
        type: typeToString(program, program.getTypeChecker().getTypeOfSymbol(parameter)),
      })),
      returnType: typeToString(program, program.getTypeChecker().getReturnTypeOfSignature(funcSignature)),
    });
  }

  //
  // Extract properties from propsType
  //

  const properties: Array<ComponentProperty> = [];
  const regions: Array<ComponentRegion> = [];
  const events: Array<EventHandler> = [];

  const interfaceProps = propsType.getProperties();
  assert.ok(interfaceProps, `No properties found for ${propsName}`);

  for (const prop of sortSymbolsByName(interfaceProps)) {
    const declaration = prop.getDeclarations()?.[0];
    assert.ok(declaration && ts.isPropertySignature(declaration) && declaration.type);

    if (declaration.type.getText() === 'React.ReactNode' || declaration.type.getText() === 'ReactNode') {
      // If a property accepts React.ReactNode, it's treated as a slot.
      const visualRefreshTag = prop.getJsDocTags().find(tag => tag.name === 'visualrefresh');
      regions.push({
        name: prop.getName(),
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        displayName: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'displayname')?.text),
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        visualRefreshTag: visualRefreshTag ? joinSymbolDisplayParts(visualRefreshTag.text) ?? '' : undefined,
        i18nTag: prop.getJsDocTags().some(tag => tag.name === 'i18n') || undefined,
        inlineType: undefined,
        isDefault: prop.getName() === 'children',
      });
    } else if (EVENT_HANDLER_REGEX.test(prop.getName())) {
      // Event handler types are expected to be either CancelableEventHandler<T> or NonCancelableEventHandler<T>.
      // The <T> is what's documented in detailType and detailInlineType.
      let detailType: ts.Type | undefined;
      let detailTypeString: string | undefined;
      let isCancelable = false;

      if (ts.isTypeReferenceNode(declaration.type)) {
        const detailTypeNode = declaration.type.typeArguments?.[0];
        if (detailTypeNode) {
          detailType = program.getTypeChecker().getTypeFromTypeNode(detailTypeNode);
          detailTypeString = nodeToString(sourceFile, detailTypeNode, ts.EmitHint.Unspecified);
        }
        isCancelable = declaration.type.typeName.getText() !== 'NonCancelableEventHandler';
      }

      events.push({
        name: prop.getName(),
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        cancelable: isCancelable,
        detailType: detailTypeString,
        detailInlineType:
          detailType && detailTypeString ? buildInlineTypeDefinition(program, detailTypeString, detailType) : undefined,
      });
    } else {
      const propType = program.getTypeChecker().getTypeFromTypeNode(declaration.type);
      const typeString = typeToString(program, propType);
      const visualRefreshTag = prop.getJsDocTags().find(tag => tag.name === 'visualrefresh');

      const property: ComponentProperty = {
        name: prop.getName(),
        type: typeString,
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        visualRefreshTag: visualRefreshTag ? joinSymbolDisplayParts(visualRefreshTag.text) ?? '' : undefined,
        i18nTag: prop.getJsDocTags().some(tag => tag.name === 'i18n') || undefined,
        optional: !!(prop.getFlags() & ts.SymbolFlags.Optional) || isUnionWithUndefined(propType),
        defaultValue: defaultValues.get(prop.getName()),
      };

      if (ts.isTypeReferenceNode(declaration.type)) {
        // This is a named type (ThingProps.Item). Generally, we expect all
        // property types to be defined this way, for developer convenience.
        property.inlineType = buildInlineTypeDefinition(program, typeString, propType);

        // For compatibility, a union of strings should always be just "string".
        if (isUnionOfStrings(propType)) {
          property.type = 'string';
        }
      } else if (isInlineUnionOfStrings(declaration.type)) {
        // This is a union type, but is defined inline.
        // The previous documenter documented this as an inline type with an empty name.
        property.type = 'string';
        property.inlineType = {
          name: '',
          type: 'union',
          values: declaration.type.types.map(t => ((t as ts.LiteralTypeNode).literal as ts.StringLiteral).text).sort(),
        };
      }

      properties.push(property);
    }
  }

  return {
    name: componentName,
    version: componentVersion,
    description: componentDoc,
    releaseStatus: componentBeta ? 'beta' : 'stable',
    properties,
    regions,
    events,
    functions,
  };
}

function getComponentFunctionNode(program: ts.Program, startNode: ts.Node): ts.SignatureDeclaration {
  let node: ts.Node | undefined = startNode;

  if (ts.isExportAssignment(node)) {
    // This is an "export default Abc" statement. This gets the node where "Abc" is declared.
    node = program.getTypeChecker().getSymbolAtLocation(node.expression)?.valueDeclaration;
    assert.ok(node);
  }

  if (ts.isVariableDeclaration(node)) {
    // This is a "const Abc = ..." statement.
    // For simplicity, we require that a variable be declared and initialized together.
    // NOTE: If needed, we'll need to pull documentation from here, not the function node.
    assert.ok(node.initializer);
    node = node.initializer;
  }

  if (ts.isAsExpression(node)) {
    // This is a "value as Type" statement.
    // NOTE: For now, we discard the type information and use the naming convention. Maybe we shouldn't?
    node = node.expression;
  }

  if (ts.isCallExpression(node)) {
    // Function call, very likely React.memo or React.forwardRef.
    // NOTE: Actually validate that this is a known/expected function?
    assert.ok(node.arguments[0]);
    node = node.arguments[0];
  }

  assert.ok(ts.isFunctionLike(node), 'Unexpected default export ts.SyntaxKind: ' + node.kind);
  return node;
}

function buildInlineTypeDefinition(program: ts.Program, name: string, type: ts.Type): InlineTypeDefinition | undefined {
  // If it's a "global" type (e.g. HTMLElement), we don't want to turn it
  // into an inline type. So we keep a blocklist for now. There may be a
  // better way to do this.
  if (PASSTHROUGH_TYPE_NAMES.includes(name)) {
    return undefined;
  }

  if (type.isUnion()) {
    return {
      name,
      type: 'union',
      values: type.types.map(t => typeToString(program, t, true)).sort(),
    };
  }

  // If a type has a call signature (even if it's defined as an object),
  // we treat it as a "function". We have no use for callable types that
  // also have properties.
  const callSignatures = type.getCallSignatures();
  if (callSignatures.length > 0) {
    const signature = callSignatures[0];
    return {
      name,
      type: 'function',
      parameters: signature.parameters.map(parameter => {
        const declaration = parameter.getDeclarations()?.[0];
        assert.ok(declaration && ts.isParameter(declaration) && declaration.type);
        return {
          name: parameter.getName(),
          type: typeToString(program, program.getTypeChecker().getTypeFromTypeNode(declaration.type)),
        };
      }),
      returnType: typeToString(program, program.getTypeChecker().getReturnTypeOfSignature(signature)),
    };
  }

  // We can treat non-union, non-callable types as object types (or types that
  // resolve down to object types). One exception is arrays, which are technically
  // object types, but we don't want JS array methods (like sort) to end up in here.
  const hasIndexSignatures = !!type.getNumberIndexType() || !!type.getStringIndexType();
  if (!type.isLiteral() && !hasIndexSignatures && type.getProperties().length > 0) {
    return {
      name,
      type: 'object',
      properties: sortSymbolsByName(type.getProperties()).map(property => {
        const declaration = property.getDeclarations()?.[0];
        assert.ok(declaration);

        if (ts.isPropertySignature(declaration)) {
          assert.ok(declaration.type);
          const type = program.getTypeChecker().getTypeFromTypeNode(declaration.type);
          return {
            name: property.getName(),
            optional: !!(property.flags & ts.SymbolFlags.Optional) || isUnionWithUndefined(type),
            type: typeToString(program, type),
          };
        }

        if (ts.isMethodSignature(declaration)) {
          return {
            name: property.getName(),
            optional: !!(property.flags & ts.SymbolFlags.Optional) || isUnionWithUndefined(type),
            type: signatureDeclarationToString(program, declaration),
          };
        }

        assert.fail(`Unexpected declaration kind: ${declaration.kind}`);
      }),
    };
  }
}

function isInlineUnionOfStrings(type: ts.TypeNode): type is ts.UnionTypeNode {
  return (
    ts.isUnionTypeNode(type) &&
    type.types.every(subType => ts.isLiteralTypeNode(subType) && ts.isStringLiteral(subType.literal))
  );
}

function isUnionOfStrings(type: ts.Type): type is ts.UnionType {
  return type.isUnion() && type.types.every(subType => subType.isStringLiteral());
}

function isUnionWithUndefined(type: ts.Type): type is ts.UnionType {
  return type.isUnion() && type.types.length === 2 && type.types.some(t => t.getFlags() & ts.TypeFlags.Undefined);
}

function stripUnionWithUndefined(type: ts.Type): ts.Type {
  if (isUnionWithUndefined(type)) {
    return type.types.find(subtype => (subtype.getFlags() & ts.TypeFlags.Undefined) === 0)!;
  }
  return type;
}
