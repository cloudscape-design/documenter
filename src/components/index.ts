// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { pascalCase } from 'change-case';
import pathe from 'pathe';
import { matcher } from 'micromatch';

import { buildComponentDefinition } from './component-definition';
import { extractDefaultValues, extractExports, extractFunctions, extractProps } from './extractor';
import type { ComponentDefinition } from './interfaces';
import { bootstrapTypescriptProject } from '../bootstrap/typescript';
import { extractDeclaration, getDescription } from './type-utils';

function componentNameFromPath(componentPath: string) {
  const dashCaseName = pathe.basename(pathe.dirname(componentPath));
  return { dashCaseName, name: pascalCase(dashCaseName) };
}

export interface DocumenterOptions {
  tsconfigPath: string;
  publicFilesGlob: string;
  extraExports?: Record<string, Array<string>>;
}

export function documentComponents(options: DocumenterOptions): Array<ComponentDefinition> {
  const program = bootstrapTypescriptProject(options.tsconfigPath);
  const checker = program.getTypeChecker();

  const isMatch = matcher(pathe.resolve(options.publicFilesGlob));

  return program
    .getSourceFiles()
    .filter(file => isMatch(file.fileName))
    .map(sourceFile => {
      const moduleSymbol = checker.getSymbolAtLocation(sourceFile);
      const { name, dashCaseName } = componentNameFromPath(sourceFile.fileName);

      // istanbul ignore next
      if (!moduleSymbol) {
        throw new Error(`Unable to resolve module: ${sourceFile.fileName}`);
      }
      const exportSymbols = checker.getExportsOfModule(moduleSymbol);
      const { propsSymbol, componentSymbol } = extractExports(
        name,
        exportSymbols,
        checker,
        options?.extraExports ?? {}
      );
      const props = extractProps(propsSymbol, checker);
      const functions = extractFunctions(propsSymbol, checker);
      const defaultValues = extractDefaultValues(componentSymbol, checker);
      const componentDescription = getDescription(
        componentSymbol.getDocumentationComment(checker),
        extractDeclaration(componentSymbol)
      );

      return buildComponentDefinition(
        name,
        dashCaseName,
        props,
        functions,
        defaultValues,
        componentDescription,
        checker
      );
    });
}
