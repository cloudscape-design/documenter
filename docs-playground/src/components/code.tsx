// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';
import Link from '@cloudscape-design/components/link';

import './code.css';

export function Code({ children }: { children: React.ReactNode }) {
  return <pre className="code">{children}</pre>;
}

export function CodeToken({
  children,
  kind,
  options,
  onNavigate,
}: {
  children: React.ReactNode;
  kind: string;
  options: any;
  onNavigate: (path: string[]) => void;
}) {
  switch (kind) {
    case 'whitespace':
      return (
        <>
          <Token kind="whitespace">{children}</Token>
          <wbr />
        </>
      );
    case 'name':
      return options?.path ? (
        <Link onFollow={() => onNavigate(options?.path)} className="code-token-link">
          <Token kind="name">{children}</Token>
        </Link>
      ) : (
        <Token kind={kind}>{children}</Token>
      );
    default:
      return <Token kind={kind}>{children}</Token>;
  }
}

function Token({ children, kind }: { children: React.ReactNode; kind: string }) {
  return <span className={`code-token-${kind}`}>{children}</span>;
}
