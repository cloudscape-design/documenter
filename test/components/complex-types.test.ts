// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ComponentDefinition } from '../../src/components/interfaces';
import { buildProject } from './test-helpers';

let sideNavigation: ComponentDefinition;
let table: ComponentDefinition;

beforeAll(() => {
  const result = buildProject('complex-types');
  expect(result).toHaveLength(2);

  [sideNavigation, table] = result;
});

test('should only have expected properties, regions and events', () => {
  expect(table.properties.map(prop => prop.name)).toEqual(['ariaLabels', 'columns', 'filteringFn', 'items', 'trackBy']);
  expect(table.events.map(prop => prop.name)).toEqual(['onWidthChange']);
  expect(table.regions.map(prop => prop.name)).toEqual(['header']);
});

test('should have correct region definition', () => {
  const header = table.regions[0];
  expect(header).toEqual({
    name: 'header',
    displayName: undefined,
    isDefault: false,
    description: undefined,
  });
});

test('should have correct property types', () => {
  expect(table.properties).toEqual([
    {
      name: 'ariaLabels',
      type: 'TableProps.AriaLabels<T>',
      optional: true,
      defaultValue: undefined,
      description: undefined,
      inlineType: {
        name: 'TableProps.AriaLabels',
        type: 'object',
        properties: [
          {
            name: 'allItemsSelectionLabel',
            optional: true,
            type: '(data: TableProps.SelectionState<T>) => string',
          },
        ],
      },
    },
    {
      name: 'columns',
      type: 'Array<TableProps.TableColumn<T>>',
      description: 'Testing array literal',
      inlineType: undefined,
      optional: false,
      defaultValue: undefined,
    },
    {
      name: 'filteringFn',
      type: 'TableProps.FilteringFunction<T>',
      inlineType: {
        type: 'function',
        name: 'TableProps.FilteringFunction',
        returnType: 'boolean',
        parameters: [
          {
            name: 'item',
            type: 'T',
          },
        ],
      },
      optional: true,
      defaultValue: undefined,
      description: undefined,
    },
    {
      name: 'items',
      description: 'Testing ReadonlyArray',
      type: 'ReadonlyArray<TableProps.Item>',
      defaultValue: undefined,
      inlineType: undefined,
      optional: false,
    },
    {
      name: 'trackBy',
      type: 'TableProps.TrackBy<T>',
      inlineType: {
        type: 'union',
        name: 'TableProps.TrackBy',
        values: ['string', '(item: T) => boolean'],
      },
      optional: true,
      defaultValue: undefined,
      description: undefined,
    },
  ]);
});

test('should have correct detail type in the event', () => {
  const event = table.events.find(prop => prop.name === 'onWidthChange');
  expect(event).toEqual({
    name: 'onWidthChange',
    detailType: 'TableProps.ColumnWidthsChangeDetail',
    detailInlineType: {
      type: 'object',
      name: 'TableProps.ColumnWidthsChangeDetail',
      properties: [
        {
          name: 'widths',
          optional: false,
          type: 'Array<number>',
        },
      ],
    },
    description: undefined,
    cancelable: false,
  });
});

test('should properly display string union types', () => {
  const eventDetail = sideNavigation.events.find(def => def.name === 'onFollow');
  expect(eventDetail?.detailType).toEqual('SideNavigationProps.FollowDetail');
  expect(eventDetail?.detailInlineType).toEqual({
    type: 'object',
    name: 'SideNavigationProps.FollowDetail',
    properties: [
      {
        name: 'href',
        optional: false,
        type: 'string',
      },
      {
        name: 'type',
        optional: true,
        type: '"expandable-link-group" | "link" | "link-group"',
      },
    ],
  });
});
