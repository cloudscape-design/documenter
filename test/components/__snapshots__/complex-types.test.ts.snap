// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should print long inline types 1`] = `
[
  {
    "analyticsTag": undefined,
    "defaultValue": undefined,
    "deprecatedTag": undefined,
    "description": undefined,
    "i18nTag": undefined,
    "inlineType": {
      "name": "ButtonProps.Style",
      "properties": [
        {
          "inlineType": {
            "name": "object",
            "properties": [
              {
                "inlineType": {
                  "name": "object",
                  "properties": [
                    {
                      "name": "active",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "default",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "disabled",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "hover",
                      "optional": true,
                      "type": "string",
                    },
                  ],
                  "type": "object",
                },
                "name": "background",
                "optional": true,
                "type": "{ active?: string | undefined; default?: string | undefined; disabled?: string | undefined; hover?: string | undefined; }",
              },
              {
                "inlineType": {
                  "name": "object",
                  "properties": [
                    {
                      "name": "active",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "default",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "disabled",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "hover",
                      "optional": true,
                      "type": "string",
                    },
                  ],
                  "type": "object",
                },
                "name": "borderColor",
                "optional": true,
                "type": "{ active?: string | undefined; default?: string | undefined; disabled?: string | undefined; hover?: string | undefined; }",
              },
              {
                "inlineType": {
                  "name": "object",
                  "properties": [
                    {
                      "name": "active",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "default",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "disabled",
                      "optional": true,
                      "type": "string",
                    },
                    {
                      "name": "hover",
                      "optional": true,
                      "type": "string",
                    },
                  ],
                  "type": "object",
                },
                "name": "color",
                "optional": true,
                "type": "{ active?: string | undefined; default?: string | undefined; disabled?: string | undefined; hover?: string | undefined; }",
              },
            ],
            "type": "object",
          },
          "name": "root",
          "optional": true,
          "type": "{ background?: { active?: string | undefined; default?: string | undefined; disabled?: string | undefined; hover?: string | undefined; } | undefined; color?: { active?: string | undefined; default?: string | undefined; disabled?: string | undefined; hover?: string | undefined; } | undefined; borderColor?: { ...; } |...",
        },
      ],
      "type": "object",
    },
    "name": "style",
    "optional": false,
    "systemTags": undefined,
    "type": "ButtonProps.Style",
    "visualRefreshTag": undefined,
  },
]
`;
